File: problems/012/012.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* Copyright 2021 Melwyn Francis Carlo */
/* File Reference: http://www.naturalnumbers.org/primes.html */

int main()
{
    FILE *fp;
    char prime_num[10];
    char* filename = "problems/003/PrimeNumbers_Upto_1000000";

    int n;
    int count = 0;
    int divisors_n = 1;

    while (divisors_n <= 500)
    {
        n = 1001 + (1000 * count);
        int n_max = n + 999;
        int divisors_arr[2][501];

        while (n <= n_max)
        {
            int n_sum = n * (n + 1) / 2;

            memset(divisors_arr, 0, sizeof(divisors_arr));

            while (n_sum != 1)
            {
                fp = fopen(filename, "r");

                while (fgets(prime_num, 10, fp) != NULL)
                {
                    if ( (n_sum % atoi(prime_num)) == 0 )
                    {
                        int index = -1;

                        for (int i = 1; i <= divisors_arr[0][0]; i++)
                        {
                            if (divisors_arr[1][i] == atoi(prime_num))
                            {
                                index = i;
                                break;
                            }
                        }

                        if (index == -1)
                        {
                            divisors_arr[0][0]++;
                            divisors_arr[0][divisors_arr[0][0]]++;
                            divisors_arr[1][divisors_arr[0][0]] = atoi(prime_num);
                        }
                        else
                        {
                            divisors_arr[0][index]++;
                        }

                        n_sum /= atoi(prime_num);
                        break;
                    }
                }

                fclose(fp);
            }

            divisors_n = 1;
            for (int i = 1; i <= divisors_arr[0][0]; i++)
            {
                divisors_n *= (divisors_arr[0][i] + 1);
            }

            n++;

            if (divisors_n > 500)
                break;
        }

        count++;
    }

    n--;

    printf("%d\n", n * (n + 1) / 2);

    return 0;
}

File: problems/012/012.f90
PROGRAM F012

    ! Copyright 2021 Melwyn Francis Carlo
    ! FILE REFERENCE: http://www.naturalnumbers.org/primes.html

    IMPLICIT NONE

    INTEGER :: PRIME_NUM
    CHARACTER (LEN=50), PARAMETER :: FILENAME = "problems/003/PrimeNumbers_Upto_1000000"

    INTEGER :: N          = 1 ! Just a dummy value.
    INTEGER :: COUNT_VAL  = 0
    INTEGER :: DIVISORS_N = 1

    INTEGER :: I, INDEX_VAL, N_MAX, N_SUM

    INTEGER, DIMENSION(2, 0:500) :: DIVISORS_ARR

    DO WHILE (DIVISORS_N <= 500)

        N     = 1001 + (1000 * COUNT_VAL)
        N_MAX = N + 999

        DO WHILE (N <= N_MAX)

            DIVISORS_ARR = 0
            N_SUM        = N * (N + 1) / 2

            DO WHILE (N_SUM /= 1)

                OPEN (2, FILE = FILENAME, STATUS = 'OLD')

                DO
                    READ(2, *, END=10) PRIME_NUM

                    IF (MOD(N_SUM, PRIME_NUM) == 0) THEN

                        INDEX_VAL = -1

                        DO I = 1, DIVISORS_ARR(1, 0)
                            IF (DIVISORS_ARR(2, I) == PRIME_NUM) THEN
                                INDEX_VAL = I
                                EXIT
                            END IF
                        END DO

                        IF (INDEX_VAL == -1) THEN
                            DIVISORS_ARR(1, 0) = DIVISORS_ARR(1, 0) + 1

                            DIVISORS_ARR(1, DIVISORS_ARR(1, 0))                &
                                = DIVISORS_ARR(1, DIVISORS_ARR(1, 0)) + 1

                            DIVISORS_ARR(2, DIVISORS_ARR(1, 0)) = PRIME_NUM
                        ELSE
                            DIVISORS_ARR(1, INDEX_VAL)                         &
                                = DIVISORS_ARR(1, INDEX_VAL) + 1
                        END IF

                        N_SUM = N_SUM / PRIME_NUM
                        EXIT

                    END IF

                END DO

                10 CLOSE(2)

            END DO

            DIVISORS_N = 1
            DO I = 1, DIVISORS_ARR(1, 0)
                DIVISORS_N = DIVISORS_N * (DIVISORS_ARR(1, I) + 1)
            END DO

            N = N + 1

            IF (DIVISORS_N > 500) EXIT

        END DO

        COUNT_VAL = COUNT_VAL + 1

    END DO

    N = N - 1

    PRINT ('(I0)'), N * (N + 1) / 2

END PROGRAM F012
